pkg/apiserver/compression_io.go

```diff
@@ -0,0 +1,61 @@
+package apiserver
+
+import (
+	"compress/gzip"
+	"io"
+
+	"github.com/gin-gonic/gin"
+)
+
+type countingBodyWriter struct {
+	gin.ResponseWriter
+	writtenUncompressedBytes int
+}
+
+func (r *countingBodyWriter) WriteString(s string) (n int, err error) {
+	writtenBytes, err := r.ResponseWriter.WriteString(s)
+	r.writtenUncompressedBytes += writtenBytes
+
+	return writtenBytes, err
+}
+
+func (r *countingBodyWriter) Write(p []byte) (n int, err error) {
+	writtenBytes, err := r.ResponseWriter.Write(p)
+	r.writtenUncompressedBytes += writtenBytes
+
+	return writtenBytes, err
+}
+
+type countingBodyReader struct {
+	io.ReadCloser
+	readCompressedBytes int
+}
+
+func (r *countingBodyReader) Read(p []byte) (int, error) {
+	readBytes, err := r.ReadCloser.Read(p)
+	r.readCompressedBytes += readBytes
+
+	return readBytes, err
+}
+
+type gzipBodyReader struct {
```

Comment:
i would have liked to see factory functions for these structs and them also being used where the structs are currently crafted manually.

Also i would've liked if this pr was split up into the functional bits it consists of.